internal/test_unit/python/unit.py

```diff
@@ -0,0 +1,371 @@
+    callchk : bool, default False
+        if true, activate valgrind with callcheck
+    """
+  
+    # initialize logger
+    logger = logging.getLogger(__name__)
+    logger.setLevel(logging.DEBUG)
+
+    # create/add console handler
+    ch = logging.StreamHandler()
+    ch.setLevel(logging.INFO)
+    logger.addHandler(ch)
+
+    # create/add file handler
+    if file_log and not cmd_only:
+        fh = logging.FileHandler(file_log)
+        fh.setLevel(logging.DEBUG)
+        logger.addHandler(fh)
+   
+    # parse xml file
+    try:
+        test_root = ET.parse(test_xml)
+    except Exception as e:
+        logger.exception(f"ERROR: Unable to parse xml from {test_xml}")
+        raise
+
+    # parse the children of the met_test element
+    if test_root.getroot().tag != 'met_test':
+        logger.error(f"ERROR: unexpected top-level element. Expected 'met_test', got '{test_root.tag}'")
+        sys.exit(1)
+    else:
+        # read test_dir
+        try:
+            test_dir = test_root.find('test_dir').text
+            mgnc = repl_env(test_dir + '/bin/mgnc.sh')
+            mpnc = repl_env(test_dir + '/bin/mpnc.sh')
+        except Exception as e:
+            logger.WARNING(f"WARNING: unable to read test_dir from {test_xml}")
+            pass
+
+        tests = build_tests(test_root)
+
+    # determine the max length of the test names
+    #   not used, unless format of test result display is changed
+    name_wid = max([len(test['name']) for test in tests])
+    
+    VALGRIND_OPT_MEM ="--leak-check=full --show-leak-kinds=all --error-limit=no -v"
+    VALGRIND_OPT_CALL ="--tool=callgrind --dump-instr=yes --simulate-cache=yes --collect-jumps=yes"
+
+    # run each test
+    for test in tests:
+    #   # print the test name ... may want to change this to only if cmd_only=False
+        logger.debug("\n")
+        logger.info(f"TEST: {test['name']}")
+
+    #   # prepare the output space
+        output_keys = [key for key in test.keys() if key.startswith('out_')]
+        outputs = [output for key in output_keys for output in test[key]]
+        for output in outputs:
+            try:
+                Path(output).unlink()
+            except FileNotFoundError:
+                pass
+            except Exception as e:
+                logger.exception()
+                raise
+            output_dir = Path(output).parent
+            output_dir.mkdir(parents=True, exist_ok=True)     #should error/warning be raised if dir already exists? 
+
+    #   # set the test environment variables
+        set_envs = []
+        if 'env' in test.keys():
+            for key, val in sorted(test['env'].items()):
+                os.environ[key] = val
+                set_cmd = f"export {key}={val}"
+                logger.debug(set_cmd)
+                set_envs.append(set_cmd)
+
+    #   # build the text command
+        cmd = test['exec'] + test['param']
+        cmd = re.sub('[ \n\t]+$', '', cmd)  # not sure this is doing what it should;
```

Comment:
Is this just removing whitespace from the end of the command? If so you can use `.rstrip()` to remove whitespace from the end of `.strip()` to remove whitespace from the beginning and end.
```suggestion
        cmd = cmd.strip()
```

Reply:
maybe? i just copied that regex string from the perl script and changed to python syntax. i actually don't think it's needed at all... i ran a few tests without it without issue, but i changed it to .strip() per your suggestion and left that in for consistency. it's in my latest commit.